csrc/attention/attention_kernels.cu

```diff
@@ -297,12 +300,30 @@ __device__ void paged_attention_kernel(
       qk += (alibi_slope != 0) ? alibi_slope * (token_idx - seq_len + 1) : 0;
 
       if (thread_group_offset == 0) {
+        // The block mask marks a slot as true if it is active or false
+        // otherwise, for an inactive slot we assign -FLT_MAX to avoid
+        // constribution to the sumexp softmax normalizer; essentially this is
+        // also setting the attention score to -inf
+        const bool is_active =
+            block_masks[seq_idx * BLOCK_SIZE * max_num_blocks_per_seq +
+                        token_idx - start_token_idx];
+        qk = is_active ? qk : -FLT_MAX;
+
```

Comment:
This part does not seem to affect the softmax correctly.

> Quote [vLLM](https://docs.vllm.ai/en/latest/dev/kernel/paged_attention.html#qk-max-and-logits): We can compare and collect the `qk_max` for all `qk`s that are calculated by current thread group.

Maybe `qk` being updated only when `thread_group_offset == 0` does not correctly affect `qk_max`? I'm not really sure how the `thread_group_offset` thing works here though - this I can try tomorrow. Also is the slot is inactive, maybe we can skip many computations to improve performance but that's definitely after we figure out a "correct" solution first.

Reply:
Well actually I don't think there's a big problem here. Yesterday I see that the attention layer output do not show zeros, but on second throughts it SHOULD NOT. The attention weights should be zero (i.e., the softmax), but the attention output should not (after multiplying with V, it's that the masked position does not contribute to the final output, but not that it makes it zero). Maybe see https://gmongaras.medium.com/how-do-self-attention-masks-work-72ed9382510f this post for how attention masks work mathematically. So, as long as we confirm QK is negative infinity we should be good (I hope).